### PR TITLE
feat: HIFLD topology

### PR DESCRIPTION
**Pull request by terrywqf**
### Purpose
Merging the python modules of the parser to generate topology based on Homeland Infra Foundation-Level (HIFLD) Data, covering transmission network, power plants, demand profiles, renewables profiles that can be used by REISE to run optimization.

### What the code is doing

- Parse the raw data from HIFLD to generate topology, plant and transmission parameters.
- Parse the raw data from EIA etc. to populate the key operation parameters of the power plants.
- Parse the population raw data to generate the load distribution. 
- Divide the topology into Western, Eastern and Texas
- Add virtual lines and decompose power plants to solve the infeasibility issue

### Testing
Unit tests
Manual tests to generate the required csv to integrate with REISE etc.

### Where to look
- All the codes inside hiflddata are the core logics to parse raw data and generate csv for the REISE optimization
- All the codes inside HIFLD_Profile_APP are about integrating the demand, hydro, wind etc. profiles into HIFLD topology
#### What files are most important?
- data_trans.py 
- plant_agg.py

### Time estimate
1 week to review all the codes. 1.5 weeks to address the review comments.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -24,3 +24,5 @@ netCDF4 = "==1.5.4"
 powersimdata = {editable = true, path = "../PowerSimData"}
 nrel-pysam = "~=2.2"
 pyproj = "==2.6.1.post1"
+networkx = "==2.5"
+haversine = "==2.3.0"
```
Comment (rouille) [inline Pipfile:28]: No need to install this package. You can use the `haversine` function in the `powersimdata.utility.distance` module

File: prereise/gather/HIFLD_Profile_APP/Demand/demand_divide.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/Demand/demand_divide.py:4]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/Demand/demand_trans.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/Demand/demand_trans.py:4]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/HIFLD_Case_for_Profile_Input/solardata.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# from getpass import getpass
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/HIFLD_Case_for_Profile_Input/solardata.py:2]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/Solar_hydro_wind/solardata_nsrdb_sam.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# from getpass import getpass
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/Solar_hydro_wind/solardata_nsrdb_sam.py:2]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/src/hydro_plant_cross_validation.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
+# coding: utf-8
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/src/hydro_plant_cross_validation.py:3]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/src/solar_plant_cross_validation.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+#!/usr/bin/env python
+# coding: utf-8
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/src/solar_plant_cross_validation.py:3]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/src/tamu_data_devide.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/src/tamu_data_devide.py:4]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/src/wind_plant_cross_validation.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
+#!/usr/bin/env python
+# coding: utf-8
+
```
Comment (rouille) [inline prereise/gather/HIFLD_Profile_APP/src/wind_plant_cross_validation.py:3]: The above lines are unnecessary. Start with the import statement

File: prereise/gather/hiflddata/bus_island_remove.py [hunk 10]
```diff
@@ -0,0 +1,93 @@
+#!/usr/bin/env python
+# coding: utf-8
+
```
Comment (rouille) [inline prereise/gather/hiflddata/bus_island_remove.py:3]: The above lines are unnecessary. Start with the import statement

File: Pipfile (hunk 1 above)
Comment (terrywqf) [reply Pipfile:28]: Interesting to know this. Make sense to use the one from powersimdata directly.

File: prereise/gather/hiflddata/tests/test_data_trans.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
+# -*- coding: utf-8 -*-
```
Comment (rouille) [inline prereise/gather/hiflddata/tests/test_data_trans.py:1]: The above line is unnecessary. Start with the import statement

File: prereise/gather/hiflddata/tests/test_plant_agg.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
+# -*- coding: utf-8 -*-
```
Comment (rouille) [inline prereise/gather/hiflddata/tests/test_plant_agg.py:1]: The above line is unnecessary. Start with the import statement

File: prereise/gather/HIFLD_Profile_APP/Demand/demand_divide.py (hunk 2 above)
Comment (terrywqf) [reply prereise/gather/HIFLD_Profile_APP/Demand/demand_divide.py:4]: I guess Mingze used some template for every new script... @Mingze-Li128 please remove all the unnecessary comments

**Comment by rouille** [comment]
General comments:
* all files, folders functions and variables must follow the snake case naming convention: use of underscores as word separators and first letter of each word written in lowercase.
* All functions must have docstrings

**Comment by Mingze-Li128** [comment]
I think we do not need to submit this script. It only divides the 3 interconnect for testing.

获取 Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by rouille** [comment]
There is a lot of CSV files. Which ones are:
* input of your code
* products of your code

The ones that are inputs of your code (external data) would need to go through the data intake procedure defined in our [contribution guide](https://breakthrough-energy.github.io/docs/dev/contribution_guide.html#data-intake-procedure)

The ones that are produced by your code should not be comitted. You should have a script or notebook in a demo folder that chains the functions in your code that generate them.

**Comment by terrywqf** [comment]
flake8 will throw errors when var and functions are not using snake case - we have corrected all of them and that is why you see the GitHub pipeline is green.
For files, folders, @Mingze-Li128 please make sure we also follow this guideline.
> * all files, folders functions and variables must follow the snake case naming convention: use of underscores as word separators and first letter of each word written in lowercase.

This is already done but we can triple check
> * All functions must have docstrings

File: prereise/gather/Readme.txt [hunk 13]
```diff
@@ -0,0 +1,88 @@
+
```
Comment (danielolsen) [inline prereise/gather/Readme.txt:1]: I don't think this file should be included, it looks like a duplicate of `prereise/gather/hiflddata/Readme.txt`

**Comment by danielolsen** [comment]
What is the purpose of the separate folders `hiflddata` and `HIFLD_Profile_APP`?

If a user wants to start from all the external inputs (plus any information that we learned in iterative exploration about which lines need to be added where, and which generators need to be split), what is the process to get a single set of data files that can be read by PowerSimData? Like @rouille suggested, a script demo-ing this would be helpful.

File: prereise/gather/Readme.txt (hunk 13 above)
Comment (terrywqf) [reply prereise/gather/Readme.txt:1]: correct. that is some cleanup missed. will move it and double check the want and unwanted...

**Comment by Mingze-Li128** [comment]
> What is the purpose of the separate folders `hiflddata` and `HIFLD_Profile_APP`?
> 
> If a user wants to start from all the external inputs (plus any information that we learned in iterative exploration about which lines need to be added where, and which generators need to be split), what is the process to get a single set of data files that can be read by PowerSimData? Like @rouille suggested, a script demo-ing this would be helpful.

`HIFLD_Profile_APP` is used to transform the format of demand.csv, wind.csv, hydro.csv and solar.csv, to make them could be read by `hiflddata` correctly. For example, the origin demand.csv has 76 columns, represent 76 zones. But in our project, we only have 52 zones, so we need to transform the 76 columns in demand.csv to 52 columns.

We need this to ensure that the project in `hiflddata` could load the demand.csv and run independently. But I think you have your own way to handle this situation. So we're not sure if we should keep this.

**Comment by rouille** [comment]
> flake8 will throw errors when var and functions are not using snake case - we have corrected all of them and that is why you see the GitHub pipeline is green.
> For files, folders, @Mingze-Li128 please make sure we also follow this guideline.
> 
> > * all files, folders functions and variables must follow the snake case naming convention: use of underscores as word separators and first letter of each word written in lowercase.
> 
> This is already done but we can triple check
> 
> > * All functions must have docstrings

Not everything is covered by flake8. For example:
**prereise/gather/HIFLD_Profile_APP/Demand/demand_divide.py** should be renamed to **prereise/gather/hifld_profile_app/demand/demand_divide.py**

**Comment by rouille** [comment]
Can prereise/gather/HIFLD_Profile_APP/Readme.docx be converted to an ASCII file?

File: prereise/gather/hiflddata/data_trans.py [hunk 14]
```diff
@@ -0,0 +1,951 @@
...
+        "FALLON",
+        "GARFIELD",
+        "ROOSEVELT",
+        "PHILLIPS",
+        "SHERIDAN",
+        "VALLEY",
+        "WIBAUX",
+    ]
+    sub = open("output/sub.csv", "w", newline="")
+    csv_writer = csv.writer(sub)
```
Comment (danielolsen) [inline prereise/gather/hiflddata/data_trans.py:431]: If the data is a data frame, can we just use the `to_csv` data frame method?

File: prereise/gather/hiflddata/data_trans.py [hunk 15]
```diff
@@ -0,0 +1,951 @@
...
+                row["STATE"],
+            ]
+        )
+
+    sub.close()
+
+    return re_code, sub_code
+
+
+def write_bus(clean_data, sub_code, re_code, kv_dict):
```
Comment (danielolsen) [inline prereise/gather/hiflddata/data_trans.py:495]: Can we separate the data filtering (building `missing_sub`) from the file writing (for which we can use the `to_csv` data frame methods)?

File: prereise/gather/hiflddata/data_trans.py [hunk 16]
```diff
@@ -0,0 +1,951 @@
...
+        "RICHLAND",
+        "FALLON",
+        "GARFIELD",
+        "ROOSEVELT",
+        "PHILLIPS",
+        "SHERIDAN",
+        "VALLEY",
+        "WIBAUX",
+    ]
+    sub = open("output/sub.csv", "w", newline="")
```
Comment (danielolsen) [inline prereise/gather/hiflddata/data_trans.py:430]: If we have the paths hard-coded like this, it means that the code is inflexible, and needs to be run from a specific location. Ideally, the user would be able to provide a root path and the files would be populated into this path.

**Comment by danielolsen** [comment]
I believe all functionality in this PR has been added or substituted in follow-up PRs to the `hifld` branch, therefore I'm going to close this. Thank you @LFPower, @terrywqf, and @Mingze-Li128 for your hard work on this project.

**Pull request closed without merging**